lib/elixir/pages/anti-patterns/code-anti-patterns.md

```diff
@@ -171,11 +171,126 @@ end
+
+#### Example
+
+The function `plot/1` tries to draw a graphic to represent the position of a point in a cartesian plane. This function receives a parameter of `Map` type with the point attributes, which can be a point of a 2D or 3D cartesian coordinate system. To decide if a point is 2D or 3D, this function uses dynamic access to retrieve values of the `Map` keys:
+
+```elixir
+defmodule Graphics do
+  def plot(point) do
+    #Some other code...
+
+    # Dynamic access to use point values
+    {point[:x], point[:y], point[:z]}
+  end
+end
+```
+
+```elixir
+iex> point_2d = %{x: 2, y: 3}
+%{x: 2, y: 3}
+iex> point_3d = %{x: 5, y: 6, z: nil}
+%{x: 5, y: 6, z: nil}
+iex> Graphics.plot(point_2d)
+{2, 3, nil}   # <= ambiguous return
+iex> Graphics.plot(point_3d)
+{5, 6, nil}
+```
+
+As can be seen in the example above, even when the key `:z` does not exist in the `Map` (`point_2d`), dynamic access returns the value `nil`. This return can be dangerous because of its ambiguity. It is not possible to conclude from it whether the `Map` has the key `:z` or not. If the function relies on the return value to make decisions about how to plot a point, this can be problematic and even cause errors when testing the code.
+
+#### Refactoring
+
+To remove this anti-pattern, whenever a `Map` has keys of `Atom` type, replace the dynamic access to its values per strict access. When a non-existent key is strictly accessed, Elixir raises an error immediately, allowing developers to find bugs faster. The next code illustrates the refactoring of `plot/1`, removing this anti-pattern:
+
+```elixir
+defmodule Graphics do
+  def plot(point) do
+    #Some other code...
+
+    # Strict access to use point values
+    {point.x, point.y, point.z}
+  end
+end
+```
+
+```elixir
+iex> point_2d = %{x: 2, y: 3}
+%{x: 2, y: 3}
+iex> point_3d = %{x: 5, y: 6, z: nil}
+%{x: 5, y: 6, z: nil}
+iex> Graphics.plot(point_2d)
+** (KeyError) key :z not found in: %{x: 2, y: 3} # <= explicitly warns that
+  graphic.ex:6: Graphics.plot/1                  # <= the :z key does not exist!
+iex> Graphics.plot(point_3d)
+{5, 6, nil}
+```
+
+As shown below, another alternative to refactor this anti-pattern is to replace a `Map` with a `struct` (named map). By default, structs only support strict access to values. In this way, accesses will always return clear and objective results:
+
+```elixir
+defmodule Point do
+  @enforce_keys [:x, :y]
+  defstruct [x: nil, y: nil]
+end
+```
+
+```elixir
+iex> point = %Point{x: 2, y: 3}
+%Point{x: 2, y: 3}
+iex> point.x   # <= strict access to use point values
+2
+iex> point.z   # <= trying to access a non-existent key
+** (KeyError) key :z not found in: %Point{x: 2, y: 3}
+iex> point[:x] # <= by default, struct does not support dynamic access
+** (UndefinedFunctionError) ... (Point does not implement the Access behaviour)
```

Comment:
````suggestion
As shown below, another alternative to refactor this anti-pattern is to use pattern matching:

```elixir
defmodule Graphics do
  def plot(%{x: x, y: y, z: z}) do
    #Some other code...

    # Strict access to use point values
    {x, y, z}
  end
end
```

```elixir
iex> point_2d = %{x: 2, y: 3}
%{x: 2, y: 3}
iex> point_3d = %{x: 5, y: 6, z: nil}
%{x: 5, y: 6, z: nil}
iex> Graphics.plot(point_2d)
** (FunctionClauseError)  no function clause matching in Graphics.plot/1
  graphic.ex:2: Graphics.plot/1                  # <= the :z key does not exist!
iex> Graphics.plot(point_3d)
{5, 6, nil}
```

Another alternative is to use structs. By default, structs only support static access to its fields, promoting cleaner patterns:

```elixir
defmodule Point.2D do
  @enforce_keys [:x, :y]
  defstruct [x: nil, y: nil]
end
```

```elixir
iex> point = %Point.2D{x: 2, y: 3}
%Point.2D{x: 2, y: 3}
iex> point.x   # <= strict access to use point values
2
iex> point.z   # <= trying to access a non-existent key
** (KeyError) key :z not found in: %Point{x: 2, y: 3}
iex> point[:x] # <= by default, struct does not support dynamic access
** (UndefinedFunctionError) ... (Point does not implement the Access behaviour)
```